src/Cool/CoolRequest.php

```diff
@@ -12,169 +12,77 @@
      *   The WOPI client url, or NULL on failure.
+     *
+     * @throws \Drupal\collabora_online\Exception\CoolRequestException
+     *   The client url cannot be retrieved.
      */
-    public function getWopiClientURL() {
-        $_HOST_SCHEME = isset($_SERVER['HTTPS']) ? 'https' : 'http';
-        $default_config = \Drupal::config('collabora_online.settings');
-        $wopi_client_server = $default_config->get('cool')['server'];
-        if (!$wopi_client_server) {
-            $this->error_code = 201;
-            return NULL;
-        }
-        $wopi_client_server = trim($wopi_client_server);
-
-        if (!strStartsWith($wopi_client_server, 'http')) {
-            $this->error_code = 204;
-            return NULL;
-        }
-
-        if (!strStartsWith($wopi_client_server, $_HOST_SCHEME . '://')) {
-            $this->error_code = 202;
-            return NULL;
-        }
-
-        $discovery = getDiscovery($wopi_client_server);
-        if ($discovery === FALSE) {
-            $this->error_code = 203;
-            return NULL;
-        }
+    public function getWopiClientURL(string $mimetype = 'text/plain'): string {
+        $discovery = $this->discoveryXmlEndpoint->getDiscoveryXml();
 
         $discovery_parsed = simplexml_load_string($discovery);
         if (!$discovery_parsed) {
-            $this->error_code = 102;
-            return NULL;
+            throw new CoolRequestException(
+                'The retrieved discovery.xml file is not a valid XML file.',
+                102,
+            );
         }
 
-        $this->wopi_src = strval(getWopiSrcUrl($discovery_parsed, 'text/plain')[0]);
-        if (!$this->wopi_src) {
-            $this->error_code = 103;
-            return NULL;
-        }
+        $wopi_src = $this->getWopiSrcUrl($discovery_parsed, $mimetype);
+
+        return $wopi_src;
+    }
 
-        return $this->wopi_src;
+    /**
+     * Extracts a WOPI url from the parsed discovery.xml.
+     *
+     * @param \SimpleXMLElement $discovery_parsed
+     *   Parsed contents from discovery.xml from the Collabora server.
+     *   Currently, NULL or FALSE are supported too, but lead to NULL return
+     *   value.
+     * @param string $mimetype
+     *   MIME type for which to fetch the WOPI url. E.g. 'text/plain'.
+     *
+     * @return string
+     *   WOPI url as configured for this MIME type in discovery.xml.
+     *
+     * @throws \Drupal\collabora_online\Exception\CoolRequestException
+     *   No WOPI url was found for this MIME type.
+     */
+    protected function getWopiSrcUrl(\SimpleXMLElement $discovery_parsed, string $mimetype): string {
+        $result = $discovery_parsed->xpath(sprintf('/wopi-discovery/net-zone/app[@name=\'%s\']/action', $mimetype));
+        if (!empty($result[0]['urlsrc'][0])) {
+            return (string) $result[0]['urlsrc'][0];
+        }
+        throw new CoolRequestException(
+            'The requested mime type is not handled.',
+            103,
+        );
```

Comment:
I still want to understand what these numbers are for.
I tried to keep them, but I suspect they are not useful for anything.

Reply:
Here is an explanation where the numbers come from:
https://github.com/CollaboraOnline/collabora-drupal/issues/56#issuecomment-2479859790

> > In `CoolRequest` we have these error codes 101, 201 etc.
> > Do these numbers have special meanings? Is there any external system that understands these number codes?
> 
> They come from our PHP example (I didn't write that):
> 
> https://github.com/CollaboraOnline/collabora-online-sdk-examples/blob/master/webapp/php/load.php
> 
> I also reuse them in the WordPress plugin.
> 
> We could use different values if that's better.

So, they seem not useful for anything, and we should drop them.
Instead, we can introduce dedicated exception types or even new codes (int, string or enum) when we actually need to distinguish different types of failure.

Reply:
Numbers are gone.